diff -Nur system/vold/fs.orig/Exfat.cpp system/vold/fs/Exfat.cpp
--- system/vold/fs.orig/Exfat.cpp	2018-03-25 17:48:58.350212158 +0800
+++ system/vold/fs/Exfat.cpp	2018-03-26 10:27:43.351870199 +0800
@@ -39,6 +39,7 @@
 
 static const char* kMkfsPath = "/system/bin/mkfs.exfat";
 static const char* kFsckPath = "/system/bin/fsck.exfat";
+static const char* kMountPath = "/system/bin/mount.exfat";
 
 bool IsSupported() {
     return access(kMkfsPath, X_OK) == 0
@@ -82,6 +83,22 @@
         rc = mount(c_source, c_target, CONFIG_EXFAT_DRIVER, flags, mountData);
     }
 
+    if (rc) {
+        sprintf(mountData,
+                "noatime,nodev,nosuid,dirsync,uid=%d,gid=%d,fmask=%o,dmask=%o,%s,%s",
+                ownerUid, ownerGid, permMask, permMask,
+                (executable ? "exec" : "noexec"),
+                (ro ? "ro" : "rw"));
+
+        std::vector<std::string> cmd;
+        cmd.push_back(kMountPath);
+        cmd.push_back("-o");
+        cmd.push_back(mountData);
+        cmd.push_back(c_source);
+        cmd.push_back(c_target);
+
+        rc = ForkExecvp(cmd);
+    }
     return rc;
 }
 
diff -Nur system/vold/fs.orig/Ntfs.cpp system/vold/fs/Ntfs.cpp
--- system/vold/fs.orig/Ntfs.cpp	2018-03-25 17:48:58.369211984 +0800
+++ system/vold/fs/Ntfs.cpp	2018-03-26 10:23:20.170315249 +0800
@@ -33,6 +33,7 @@
 
 static const char* kMkfsPath = "/system/bin/mkfs.ntfs";
 static const char* kFsckPath = "/system/bin/fsck.ntfs";
+static const char* kMountPathK = "/system/bin/mount";
 static const char* kMountPath = "/system/bin/mount.ntfs";
 
 bool IsSupported() {
@@ -60,6 +61,32 @@
     const char* c_source = source.c_str();
     const char* c_target = target.c_str();
 
+    {
+        sprintf(mountData,
+            "utf8,uid=%d,gid=%d,fmask=%o,dmask=%o,nodev,nosuid",
+            ownerUid, ownerGid, permMask, permMask);
+
+        if (!executable)
+            strlcat(mountData, ",noexec", sizeof(mountData));
+        if (ro)
+            strlcat(mountData, ",ro", sizeof(mountData));
+        if (remount)
+            strlcat(mountData, ",remount", sizeof(mountData));
+
+        std::vector<std::string> cmd;
+        cmd.push_back(kMountPathK);
+        cmd.push_back("-t");
+        cmd.push_back("ntfs");
+        cmd.push_back("-o");
+        cmd.push_back(mountData);
+        cmd.push_back(c_source);
+        cmd.push_back(c_target);
+
+        status_t res = ForkExecvp(cmd);
+        if (res == OK)
+            return res;
+    }
+
     sprintf(mountData,
             "utf8,uid=%d,gid=%d,fmask=%o,dmask=%o,"
             "shortname=mixed,nodev,nosuid,dirsync",
--- system/vold/Utils.cpp.orig	2018-03-25 22:50:12.568481109 +0800
+++ system/vold/Utils.cpp	2018-03-25 22:50:46.643182364 +0800
@@ -571,6 +571,7 @@
     if (supported.find(CONFIG_EXFAT_DRIVER "\n") != std::string::npos)
         supported.append("nodev\texfat\n");
 #endif
+    supported.append("fuse\texfat\n");
 
     return supported.find(fsType + "\n") != std::string::npos;
 }
