--- frameworks/base/services/core/java/com/android/server/connectivity/Tethering.java.orig	2018-02-14 06:55:15.102032989 +0800
+++ frameworks/base/services/core/java/com/android/server/connectivity/Tethering.java	2018-02-14 06:55:46.766701043 +0800
@@ -1133,6 +1133,12 @@
             return;
         }
 
+        // Hack for fix WIFI Tether on some devices
+        if(mTetherStates.get(chosenIface) == null) {
+            Log.e(TAG, "Adding non-existent Iface: " + chosenIface);
+            trackNewTetherableInterface(chosenIface, interfaceType);
+        }
+
         int result = (enable ? tether(chosenIface) : untether(chosenIface));
         if (result != ConnectivityManager.TETHER_ERROR_NO_ERROR) {
             Log.e(TAG, "unable start or stop tethering on iface " + chosenIface);
